Frameworks/include/NSOrderedSetInternal.h

```diff
@@ -0,0 +1,28 @@
+//******************************************************************************
+//
+// Copyright (c) 2016 Microsoft Corporation. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+#pragma once
+
+#import <Starboard.h>
+#import <Foundation/NSOrderedSet.h>
+#import <map>
+#import <vector>
+
+@interface NSOrderedSet () {
+@protected
+    std::vector<StrongId<id>> setContainer;
+    std::map<id, std::vector<StrongId<id>>::iterator> mapContainer;
```

Comment:
It is an odd data structure but yes I think a combination of NSArray + NSSet is the way to go. Keeping them in sync is a bit of a challenge but I don't believe Apple makes any thread safe promises about ordered set so it shouldn't be a huge burden. I see what ben is saying about an ordering on a hash table but that isn't what we want here. Hash table orderings are usually used internally for faster operations which implies an ordering according to some well defined scheme. NSOrderedSet however allows for programmatic index manipulations (insertAtIndex etc) that act like an array. A hash table with a linked list would need to be a single linked list which removes the benefit of the hash table unless I'm misunderstanding Ben's idea. For simplicity I would go with NSArray and NSSet as you have suggested.

---
In reply to: [62720222](https://github.com/Microsoft/WinObjC/pull/482#discussion_r62720222) [](ancestors = 62720222,62575598)